rfc-165-add-api-endpoints-to-transition.md

```diff
@@ -0,0 +1,40 @@
+# Add API endpoints to Transition for consumption by Bouncer
+
+## Summary
+
+This proposes that the Bouncer application should access transition data via
+API endpoints on Transition instead of via direct database access.
+
+## Problem
+
+The [Transition
+System](https://docs.publishing.service.gov.uk/manual/transition-architecture.html)
+is built to transition government websites to GOV.UK. It includes:
+- Transition - Ruby on Rails application used by admins to create mappings from
+  old URLs to pages on GOV.UK.
+- Bouncer - Rack-based application that uses the mappings created by Transition
+  and handles requests to those old domains.
+
+Currently, Bouncer retrieves the mappings written by Transition through direct
+read-only access to the database shared by the two applications.
+
+This is problematic as any code related to database access must be kept in sync
+between Transition and Bouncer.
+This includes the use of ActiveRecord; we must take the effort to keep
+ActiveRecord subclasses and dependency versions consistent.
+
+## Proposal
+
+We should add API endpoints to Transition that will enable us to retrieve the
+data required for Bouncer to function.
+
+This approach is an alternative to merging to the Bouncer codebase into
+Transition entirely. Merging the applications is a more complex task that
+involves:
+- Rewriting large parts of Bouncer as it's currently not a Rails app.
+- Handling of routes that are the same in Bouncer and Transition.
+- Dealing with infrastructure complexity - references to Bouncer are littered 
+  through the infrastructure.
```

Comment:
You've included a breakdown of some of the implementation details of the alternative option and haven't gone into as much detail about the option we're proposing. If you left that out because it didn't seem noteworthy, I just thought I should put in a vote for: the contrast between the two lists could be interesting. And they might help illustrate the 25% estimate/help readers make their own estimates.

Reply:
I'm coming at this from a place of significant ignorance about these apps, so please excuse me if what I say here doesn't make sense:

> - Rewriting large parts of Bouncer as it's currently not a Rails app.

Can you explain a bit more about why you'd need to rewrite the code from the Rack app? e.g. why couldn't you move the Rack app code into the Rails app and [set up Rails routes for the Rack app](https://guides.rubyonrails.org/routing.html#routing-to-rack-applications).
 
> - Handling of routes that are the same in Bouncer and Transition.

Could this be handled by moving the existing Transition routes under a namespace? I'm making a (possibly incorrect) assumption that there's a requirement to retain existing Bouncer routes, but existing Transition routes could more easily change...?

> - Dealing with infrastructure complexity - references to Bouncer are littered 
  through the infrastructure.

That's unfortunate. Are there any fewer references to Transition? i.e. could you merge Transition into Bouncer...?

Reply:
Hi James, thanks for your comment. We are going to raise the idea of merging Transition into Bouncer in the next team dev chat. We can hopefully get a suitable response to you then. Can I ask whether you prefer the idea of merging the apps in general, or do you simply not find the justifications to not merge them convincing?

Reply:
> We are going to raise the idea of merging Transition into Bouncer in the next team dev chat.

I really don't know whether that's sensible or not (or even possible!) - it was just a thought!

> Can I ask whether you prefer the idea of merging the apps in general, or do you simply not find the justifications for not merging them convincing?

From what little I know (massive caveat!) and with what you've explained here the two apps sound very coupled to each other, so I'd lean towards merging them.

Reply:
Hi James, We had a look at this idea and have decided against it. This is in part because of the added complexity but also because it means that Transition and Bouncer become Publishing and BE apps respectively. Figured this fits in with the work we've been doing to separate FE functionality from Whitehall. 

On the complexity front, I think it's worth noting that these aren't massive bits of GOV.UK and what we're proposing is an improvement. They can always be merged at a later date if we decide that the API approach isn't working.